File: 18110/18110.py
# **************************************************************************** #
#                                                                              #
#                                                         :::      ::::::::    #
#    18110.py                                           :+:      :+:    :+:    #
#                                                     +:+ +:+         +:+      #
#    By: jihoolee <jihoolee@student.42SEOUL.kr>     +#+  +:+       +#+         #
#                                                 +#+#+#+#+#+   +#+            #
#    Created: 2024/03/16 18:12:21 by jihoolee          #+#    #+#              #
#    Updated: 2024/03/16 21:31:00 by jihoolee         ###   ########.fr        #
#                                                                              #
# **************************************************************************** #

import sys


PERCENTAGE: int = 30
input = sys.stdin.readline


def round_half_up(number: float, ndigits: int = 0) -> int:
    multiplier: int = 10 ** ndigits
    return int(number * multiplier + 0.5) / multiplier


def get_trimmed_mean(scores: list[int], percentage: int) -> int:
    if len(scores) == 0:
        return 0

    cutoff: int = int(round_half_up(len(scores) * percentage / 200))
    result: int = 0
    for idx in range(cutoff, len(scores) - cutoff):
        result += scores[idx]
    return int(round_half_up(result / (len(scores) - 2 * cutoff)))


def main() -> None:
    scores = sorted([int(input()) for _ in range(int(input()))])
    print(get_trimmed_mean(scores, PERCENTAGE))


if __name__ == '__main__':
    main()

File: 18110/18110.cpp
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   18110.cpp                                          :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: jihoolee <jihoolee@student.42SEOUL.kr>     +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2024/03/16 16:54:38 by jihoolee          #+#    #+#             */
/*   Updated: 2024/03/16 18:10:27 by jihoolee         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include <iostream>
#include <cmath>
#include <set>

const int PERCENTAGE = 30;

int get_trimmed_mean(const std::multiset<int>& scores, int percentage) {
  if (scores.empty())
    return 0;

  int   cutoff =
          std::round(static_cast<float>(scores.size()) * percentage / 200.);
  float result = 0;
  std::multiset<int>::const_iterator  end_it = scores.cend();
  std::multiset<int>::const_iterator  begin_it = scores.cbegin();

  for (int i = 0; i < cutoff; ++i) {
    ++begin_it;
    --end_it;
  }
  for (std::multiset<int>::const_iterator it = begin_it;
        it != end_it;
        ++it)
    result += *it;

  return round(result / (scores.size() - 2 * cutoff));
}

void get_input(std::multiset<int>* s) {
  int score_cnt;
  int num;

  std::cin >> score_cnt;
  for (int i = 0; i < score_cnt; ++i) {
    std::cin >> num;
    s->insert(num);
  }
}

int main(void) {
  std::multiset<int> scores;

  get_input(&scores);
  std::cout << get_trimmed_mean(scores, PERCENTAGE);
  return 0;
}
